File: frontend/types/inventory.ts
/** Mirrors apps.inventory.models.InventoryCategory */
export type InventoryCategory = "DETERGENT" | "SOFTENER" | "PACKAGING" | "EQUIPMENT" | "OTHER";

/** Mirrors apps.inventory.models.InventoryUnit */
export type InventoryUnit = "L" | "KG" | "PCS" | "ML";

/** Mirrors apps.inventory.models.InventoryTransaction.ChangeType */
export type InventoryChangeType = "RESTOCK" | "USAGE" | "ADJUSTMENT";

/** Mirrors apps.inventory.api.serializers.inventory.InventoryItemSerializer */
export interface InventoryItem {
  id: number;
  name: string;
  category: InventoryCategory;
  unit: InventoryUnit;
  quantity: string; // DecimalField serializes as string; read-only — see InventoryItemInput
  low_stock_threshold: string;
  notes: string;
  is_low_stock: boolean;
  created_at: string;
  updated_at: string;
}

/** Payload for POST /api/v1/inventory/items/ and PATCH .../{id}/ — `quantity`
 * is intentionally not settable here; it only changes via a logged
 * InventoryTransaction (see adjustInventoryItem). */
export interface InventoryItemInput {
  name: string;
  category: InventoryCategory;
  unit: InventoryUnit;
  low_stock_threshold?: number;
  notes?: string;
}

/** Query params for GET /api/v1/inventory/items/ */
export interface ListInventoryItemsParams {
  page?: number;
  category?: InventoryCategory;
  search?: string;
}

/** Mirrors apps.inventory.api.serializers.inventory.InventoryTransactionSerializer */
export interface InventoryTransaction {
  id: number;
  item: number;
  change_type: InventoryChangeType;
  quantity_change: string;
  reason: string;
  created_by: string | null; // StringRelatedField
  created_at: string;
}

/** Payload for POST /api/v1/inventory/items/{id}/adjust/ — RESTOCK requires a
 * positive quantity_change, USAGE requires a negative one. */
export interface AdjustInventoryItemInput {
  change_type: InventoryChangeType;
  quantity_change: number;
  reason?: string;
}

File: Backend/apps/inventory/api/serializers/inventory.py
from rest_framework import serializers

from apps.inventory.models import InventoryItem, InventoryTransaction


class InventoryItemSerializer(serializers.ModelSerializer):
    is_low_stock = serializers.BooleanField(read_only=True)

    class Meta:
        model = InventoryItem
        fields = [
            "id",
            "name",
            "category",
            "unit",
            "quantity",
            "low_stock_threshold",
            "notes",
            "is_low_stock",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "quantity", "is_low_stock", "created_at", "updated_at"]
        # `quantity` is intentionally read-only here — it can only change via
        # InventoryTransaction (see InventoryItemAdjustView), so every change
        # is auditable and (if it crosses the threshold) triggers an alert.


class InventoryTransactionSerializer(serializers.ModelSerializer):
    created_by = serializers.StringRelatedField()

    class Meta:
        model = InventoryTransaction
        fields = ["id", "item", "change_type", "quantity_change", "reason", "created_by", "created_at"]
        read_only_fields = fields


class InventoryAdjustSerializer(serializers.Serializer):
    change_type = serializers.ChoiceField(choices=InventoryTransaction.ChangeType.choices)
    quantity_change = serializers.DecimalField(max_digits=10, decimal_places=2)
    reason = serializers.CharField(required=False, allow_blank=True, default="")

    def validate(self, attrs):
        if attrs["change_type"] == InventoryTransaction.ChangeType.RESTOCK and attrs["quantity_change"] <= 0:
            raise serializers.ValidationError({"quantity_change": "Restocks must be a positive amount."})
        if attrs["change_type"] == InventoryTransaction.ChangeType.USAGE and attrs["quantity_change"] >= 0:
            raise serializers.ValidationError({"quantity_change": "Usage must be a negative amount."})
        return attrs
